sdk/core/core-websockets/src/reliableConnectionClient.ts

```diff
@@ -0,0 +1,435 @@
+   */
+  retryOptions?: RetryOptions;
+  /**
+   * The identifier of the connection.
+   */
+  identifier?: string;
+}
+
+/**
+ * Creates a reliable connection client factory.
+ */
+export function createReliableConnectionClient<SendDataT, ReceiveDataT>(
+  client: ConnectionManager<SendDataT, ReceiveDataT>,
+  createOptions: CreateReliableConnectionOptions = {},
+): (opts?: ReliableConnectionOptions) => ReliableConnectionClient<SendDataT, ReceiveDataT> {
+  const { isRetryable, resolveOnUnsuccessful } = createOptions || {};
+  type WritableClient = Writable<ReliableConnectionClient<SendDataT, ReceiveDataT>>;
+  return ({
+    retryOptions: inputRetryOptions,
+    identifier,
+  }: ReliableConnectionOptions = {}): ReliableConnectionClient<SendDataT, ReceiveDataT> => {
+    const connectionId = identifier || getRandomName();
+    const retryOptions = createFullRetryOptions(inputRetryOptions);
+    let canReconnect: boolean = true;
+    function isOpenRetryable(err: unknown): boolean {
+      const res = canReconnect && (isRetryable?.(err) ?? true);
+      canReconnect = true;
+      return res;
+    }
+    let openAbortSignal: AbortSignal | undefined = undefined;
+    let openOrClosePromiseResolve: (() => void) | undefined = undefined;
+    let openOrClosePromiseReject: ((reason?: any) => void) | undefined = undefined;
+    function clearResolvers(): void {
+      openOrClosePromiseResolve = undefined;
+      openOrClosePromiseReject = undefined;
+    }
+    function rejectAndReset(reason?: any): void {
+      openOrClosePromiseReject?.(reason);
+      clearResolvers();
+    }
+    function resolveAndReset(): void {
+      openOrClosePromiseResolve?.();
+      clearResolvers();
+    }
+    const reliableConnectionClient: WritableClient = {
+      status: "disconnected",
+      open: async (openOpts: OpenOptions = {}) => {
+        switch (reliableConnectionClient.status) {
+          case "connecting":
+          case "connected": {
+            logger.info(`[${connectionId}] Client is already ${reliableConnectionClient.status}`);
+            return;
+          }
+          case "disconnecting": {
+            const errMsg = `[${connectionId}] Client is already ${reliableConnectionClient.status}`;
+            if (resolveOnUnsuccessful) {
+              logger.warning(errMsg);
+              return;
+            } else {
+              throw createError(errMsg);
+            }
+          }
+          case "disconnected": {
+            break;
+          }
+        }
+        const { abortSignal } = openOpts;
+        openAbortSignal = abortSignal ?? openAbortSignal;
+        async function retryableConnecting(
+          opOptions: AbortableOperationOptions = {},
+        ): Promise<void> {
+          logger.verbose(`[${connectionId}] Opening connection`);
+          client.open({ ...openOpts, ...opOptions });
+          /** installs dummy handlers to activate the logic in the wrappers */
+          reliableConnectionClient.onopen(() => {});
```

Comment:
not sure I understand the purpose of these empty delegates, but if we do really need them can we cache one single no-op function ref instead?